pkg/apis/discovery/types.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// EndpointSlice represents a subset of the endpoints that implement a service.
+// For a given service there may be multiple EndpointSlice objects, selected by
+// labels, which must be joined to produce the full set of endpoints.
+type EndpointSlice struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// +optional
+	metav1.ObjectMeta
+	// endpoints is a list of unique endpoint objects in this slice.
+	// +listType=atomic
+	Endpoints []Endpoint
+	// ports specifies the list of network ports exposed each endpoint in this
+	// EndpointSlice. Each port must have a unique name. When endpoints is nil,
+	// it indicates that ports are irrelevant for the address type. When this is
+	// an empty slice, it indicates that ports are relevant for the address type,
+	// but none are currently relevant. The most common example of this occurs
+	// when no pods match a Service selector and a placeholder EndpointSlice is
+	// generated.
+	// +optional
+	// +listType=atomic
+	Ports []EndpointPort
+	// addressType specifies the type of target carried by this EndpointSlice.
+	// All endpoints in this slice must be the same type.
+	// Default is IP.
+	// +optional
+	AddressType *AddressType
+}
+
+// Endpoint represents a single logical "backend" implementing a service.
+type Endpoint struct {
+	// addresses of this endpoint. The contents of this field are interpreted
+	// according to the corresponding EndpointSlice addressType field.
+	// Different consumers (e.g. kube-proxy) handle different types of addresses
+	// in the context of its own capabilities.
+	// Must contain at least one address. A given endpoint may have multiple
+	// addresses. For example, a dual-stack system might have both IPv4 and IPv6
+	// addresses.
+	// +listType=set
+	Addresses []string
```

Comment:
do we want to specify a limit to number of addresses?  100?  200?  500?

Reply:
Can we say AddressType==IP, then at most 2 Address?

And then we extend the validation based on AddressType

Reply:
No, there are people doing multi-IP stuff.  In fact, that may be a reason to change this to 

```
Addresses []EndpointAddress
```
```
type EndpointAddress struct {
    address strring
}
```

So we can later copy metadata from Pod.IPs[].

Thoughts?  It's kind of obnoxious and would add another dimension of inefficiency.  I don't know if we'll need it so we can revisit as we progress alpha.

How about a limit of 100 with validation?

Reply:
For IPs, 16 would probably be enough.  4 would probably be enough, but let's just set it big enough that nobody complains?

Reply:
I added an initial limit of 100 here.

Reply:
I added an additional limit of 100 here, happy to change to something else if it makes more sense